packages/hardhat-core/src/internal/hardhat-network/provider/BlockchainBase.ts

```
@@ -73,22 +73,32 @@ export abstract class BlockchainBase {
     }
 
     if (typeof blockHashOrNumber === "number") {
-      return this._data.getBlockByNumber(BigInt(blockHashOrNumber)) ?? null;
+      const blockByNumber = this._data.getBlockByNumber(
+        BigInt(blockHashOrNumber)
+      );
+      if (blockByNumber === undefined) {
+        throw new Error("Block not found");
+      }
+      return blockByNumber;
```

We now throw instead of returning null when the block doesn't exist.